git-hooks/commit-msg

```diff
@@ -0,0 +1,134 @@
+#!/usr/bin/env node
+
+const fs = require('fs')
+const path = require('path')
+
+/* ---------- USER FEEDBACK MESSAGES ---------- */
+
+const SYNTAX = '<type>[(<scope>)][!]: <summary>'
+
+const incorrectFormat = (txtMsg) => `
+    [ERROR] Commit message header does not match expected format.
+    ----------------------------------------
+    Given header:
+    ${txtMsg}
+
+    Expected message header format:
+    ${SYNTAX}
+`
+
+const unknownType = (type) => `
+    ERROR: Unknown commit type!
+    ----------------------------------------
+    Given <type>:
+    ${type}
+
+    Expected <type> to be one of the following:
+        ${Object.keys(TYPES).join(', ')}
+
+    Syntax:
+    ${SYNTAX}
+`
+
+const TYPES = {
+    'build': 'update/modify build system and/or dependencies',
+    'ci': 'modification of CI configuration',
+    'docs': 'documentation-only changes',
+    'feat': 'add new functionality (semver MINOR)',
+    'fix': 'patch bug/error (semver PATCH)',
+    'perf': 'code changes that improve performance',
+    'refactor': 'code changes that neither fix a bug nor add a feature',
+    'style': 'code changes that do not modify algorithms (white space, linting, etc.)',
+    'test': 'add missing tests or update existing tests',
+}
+
+/**
+ * @param string header
+ * @returns object - object with the following keys:
+ *  - isBreaking: true if header implies a breaking change with an exclamation point
+ *  - scope: matched value for the optional <scope> portion of the header
+ *  - summary: matched value for the <summary> portion of the header
+ *  - type: matched value for the <type> portion of the header
+ */
+function parseHeader (header) {
+    let pattern = /^(?<type>[^\(!]+)(\((?<scope>.+)\))?(?<isBreaking>!)?: (?<summary>.+)$/
```

Comment:
`<type>`, `<scope>`, `<isBreaking>`, `<summary>` are all named capture groups (requires modern Node) that allows me to reference the captured value by group name, rather than by group order